### DISCUSSION TITLE
Multiple requests to json file when using  res.writeHead(301, { Location: req.url })

### DISCUSSION THREAD

**Discussion by Barik85** [vercel/next.js#64383] [Help] [answered]
### Summary

I am using next: 12.2.0
I have route `/pl/levels/upper-intermediate-english` that should be redirected with 301 statusCode to `/levels/upper-intermediate-english`
For this I use following code in `getServerSideProps`:
```
export const getServerSideProps: GetServerSideProps = async ({
  params,
  req,
  res,
  locale,
  defaultLocale,
}) => {
...
  if (!availableLocales.includes(locale || '') && req.url) {
      res.writeHead(301, { Location: req.url })
      res.end()
    }
 ...
}
```
everything works fine if i go directly to the route `/pl/levels/upper-intermediate-english`.
But when it navigates on client side by `<Link href="pl/levels/upper-intermediate-english">` I got multiple requests to json `http://localhost:3000/_next/data/development/pl/levels/upper-intermediate-english.json?slug=upper-intermediate-english`
It can be around 20 requests. As a result it slow down user navigation to desired page.

Any ideas how to fix it?
<img width="1173" alt="Знімок екрана 2024-04-12 о 10 47 12" src="https://github.com/vercel/next.js/assets/31065945/c8f72865-bdd8-4540-b469-c1a9325fcb1e">

I tried to use other solution for redirecting:
```
export const getServerSideProps: GetServerSideProps = async ({
  params,
  req,
  res,
  locale,
  defaultLocale,
}) => {
...
  if (!availableLocales.includes(locale || '') && req.url) {
     return {
      redirect: {
        destination: req.url,
        permanent: true,
      },
    }
...
}
```
But it redirects user not to html, but to json file, to the route like: _next/data/ownuOVpxdLJdgcJD1xu3D/pl/levels/upper-intermediate-english.json. It looks really strange... I can't understand why it work in this way.

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.6.0: Wed Jul  5 22:21:56 PDT 2023; root:xnu-8796.141.3~6/RELEASE_X86_64
    Binaries:
      Node: 16.13.1
      npm: 8.1.2
      Yarn: 3.2.1
      pnpm: N/A
    Relevant packages:
      next: 12.2.0
      eslint-config-next: 12.2.0
      react: 17.0.2
      react-dom: 17.0.2
```

**Comment by icyJoseph** [comment, marked as answer]
Could you give it a go with, https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props#context-parameter `resolvedUrl`?

>  resolvedUrl:  A normalized version of the request URL that strips the _next/data prefix for client transitions and includes original query values.

And using the second alternative you have in the examples, which is the _idiomatic_ way I guess.

**Reply by icyJoseph** [reply]
Although the problem I guess it that, as you have it, it works, but there's no base case when the JSON data for the page is fetched from the client 🤔 I guess `req.url`, just makes it point back to itself? and just keep on redirecting for ever, until the browser decides to stop.

**Reply by Barik85** [reply]
```
if (!availableLocales.includes(locale || '')) {
    return {
      redirect: {
        destination: resolvedUrl,
        permanent: true,
      },
    }
  }
  ```
  Works like a charm! :rocket:
  @icyJoseph Thank you for so quick answer! You made my day!

**Comment by Barik85** [comment]
When I am using same aproach in `getStaticProps` it works in development mode, but during build it fails with error:
`Error: redirect can not be returned from getStaticProps during prerendering `
From the docs [https://nextjs.org/docs/pages/api-reference/functions/get-static-props#redirect](https://nextjs.org/docs/pages/api-reference/functions/get-static-props#redirect) it should work, I can assume that will work only with `getStaticPaths` + `fallback: true/blocking`.
I tried to write redirects in `next.config.js`:
```
{
    source: '/:locale/my-path',
    destination: '/my-path',
    permanent: true,
 },
```
and received infinite requests with 308 status.
`source: '/:locale(pl|fr|es)*/confusing-words',` - same result.
What suggestion can be used for this case?

**Reply by icyJoseph** [reply]
Hi, yes with `getStaticPaths` you also control what pages to pre-render per locale, in addition to other parameters. And that would be, only the locales you want to support, from next.config.js i18n, if I remember correctly. You should also include fallback either, true, or blocking as you say.

This way, during build time `getStaticProps`, is not invoked with the locales that have to be redirected, and it is only at runtime, when a request when unsupported locales arrive, that you redirect the user.

Middleware could also work. I think next.config.js has an option to include the locale when matching redirects too.

**Reply by Barik85** [reply]
I tried getStaticPaths, and got this error:
`Error: getStaticPaths is only allowed for dynamic SSG pages and was found on '/some-static-url'`
So still in searching of solution redirect with SSG...

**Reply by icyJoseph** [reply]
Right, getStaticPaths can only be used with getStaticProps side by side.

**Reply by Barik85** [reply]
So, my journey is going on.
Now I tried middleware.
```
export function middleware(request: NextRequest) {
  const { locale, pathname, origin } = request.nextUrl

  if (
    REDIRECT_BY_LOCALES_ROUTES.some((route) => pathname?.startsWith(route)) &&
    locale !== DEFAULT_LOCALE
  ) {
    return NextResponse.redirect(`${origin}${pathname}`, { status: 301 })
  }

  return NextResponse.next()
}

export const config = {
  matcher: [
    '/route1',
    '/route1/:path*',
    ...
  ],
}

export const REDIRECT_BY_LOCALES_ROUTES = [
  '/route1',
  '/route2,
  ...
]
```
Everything was good until i tried client side navigation.
When I navigated with `<Link>` component browser starts infinite request to json file with 301 response status. And do not navigate to desired page at all.
